### ISSUE TITLE
Import doesn't populate media_gallery tables

### ISSUE THREAD

**Issue by peterjaap** [magento/magento2#18553]
### Preconditions (*)

1. Magento 2.2.4, 2.2.5, 2.2.6

### Steps to reproduce (*)

1. Import data through the command line (we use Firegento_FastSimpleImport2 as a wrapper) with images

### Expected result (*)

1. Images in an imported product;

![image](https://user-images.githubusercontent.com/431360/46864530-f635e980-ce1a-11e8-8d9e-182799ad5d2b.png)

### Actual result (*)

1. No images in an imported product;

![image](https://user-images.githubusercontent.com/431360/46864543-0057e800-ce1b-11e8-9c95-151da1081fb6.png)

However, I do see an image in the grid;

![image](https://user-images.githubusercontent.com/431360/46864561-106fc780-ce1b-11e8-8032-85d3f336257d.png)

So it appears the import doesn't populate media_gallery tables, although it does fill the `base_image`, `small_image` and `thumbnail` attributes.

I tracked it down to `\Magento\CatalogImportExport\Model\Import\Product\MediaGalleryProcessor` where in the `saveMediaGallery` method this is called;

`$productId = $this->skuProcessor->getNewSku($productSku)[$this->getProductEntityLinkField()];`

Except, the skuProcessor (`\Magento\CatalogImportExport\Model\Import\Product\SkuProcessor`) doesn't hold *any* SKU's (be it either old or new ones). The newSkus field in the SkuProcessor is populated in `\Magento\CatalogImportExport\Model\Import\Product` and when I xdebug it, I do see sku's there.

It seems the skuProcessor instance in MediaGalleryProcessor is a brand new instance, and not the instance created and populated in `\Magento\CatalogImportExport\Model\Import\Product`. Although I can't find a `shared="false"` in a `di.xml` file anywhere. Sharing (singleton pattern) is the default behavior, except when an object is created through a factory. I also can't find any factory reference for the skuProcessor class. So I have no clue why it's a new instance. My only workaround for now was to actually pass along the populated skuProcessor from `\Magento\CatalogImportExport\Model\Import\Product` (`_saveProducts` -> `_saveMediaGallery`) to `\Magento\CatalogImportExport\Model\Import\Product\MediaGalleryProcessor` (`saveMediaGallery`).

Patch file here; 

```
From b97ed077aa3d08b85bfb2839e28c566c1b94aa28 Mon Sep 17 00:00:00 2001
From: peterjaap <peterjaap@elgentos.nl>
Date: Thu, 13 Sep 2018 15:06:46 +0200
Subject: [PATCH 1/1] Passed skuProcessor from Import/Product to
 MediaGalleryProcessor to insert media gallery entries correctly

Signed-off-by: peterjaap <peterjaap@elgentos.nl>
---
 Model/Import/Product.php                       | 8 +++++---
 Model/Import/Product/MediaGalleryProcessor.php | 8 ++++++--
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/Model/Import/Product.php b/Model/Import/Product.php
index 2199936..aa77afa 100644
--- a/Model/Import/Product.php
+++ b/Model/Import/Product.php
@@ -1843,7 +1843,8 @@ class Product extends \Magento\ImportExport\Model\Import\Entity\AbstractEntity
             )->_saveProductTierPrices(
                 $tierPrices
             )->_saveMediaGallery(
-                $mediaGallery
+                $mediaGallery,
+                $this->skuProcessor
             )->_saveProductAttributes(
                 $attributes
             )->updateMediaGalleryLabels(
@@ -2044,14 +2045,15 @@ class Product extends \Magento\ImportExport\Model\Import\Entity\AbstractEntity
      * Save product media gallery.
      *
      * @param array $mediaGalleryData
+     * @param null $skuProcessor
      * @return $this
      */
-    protected function _saveMediaGallery(array $mediaGalleryData)
+    protected function _saveMediaGallery(array $mediaGalleryData, $skuProcessor = null)
     {
         if (empty($mediaGalleryData)) {
             return $this;
         }
-        $this->mediaProcessor->saveMediaGallery($mediaGalleryData);
+        $this->mediaProcessor->saveMediaGallery($mediaGalleryData, $skuProcessor);
 
         return $this;
     }
diff --git a/Model/Import/Product/MediaGalleryProcessor.php b/Model/Import/Product/MediaGalleryProcessor.php
index ec7c6a1..a9c39ec 100644
--- a/Model/Import/Product/MediaGalleryProcessor.php
+++ b/Model/Import/Product/MediaGalleryProcessor.php
@@ -103,11 +103,15 @@ class MediaGalleryProcessor
     /**
      * Save product media gallery.
      *
-     * @param $mediaGalleryData
+     * @param array $mediaGalleryData
+     * @param null $skuProcessor
      * @return void
      */
-    public function saveMediaGallery(array $mediaGalleryData)
+    public function saveMediaGallery(array $mediaGalleryData, $skuProcessor = null)
     {
+        if ($skuProcessor !== null) {
+            $this->skuProcessor = $skuProcessor;
+        }
         $this->initMediaGalleryResources();
         $mediaGalleryDataGlobal = array_replace_recursive(...$mediaGalleryData);
         $imageNames = [];
-- 
2.17.1

```

**Comment by magento-engcom-team** [comment]
Hi @peterjaap. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@peterjaap do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [x] no

**Comment by peterjaap** [comment]
I can't reproduce it in a vanilla instance since I need a wrapper to programatically import the products.

**Comment by peterjaap** [comment]
Weirdly enough, when I set the `$skuProcessor` field in `MediaGalleryProcessor` and `Product` to public and use the repl to check the object reference, it **is** the same instance;

```
$ magerun2 dev:console
Magento 2.2.5 Community initialized. ✔
At the prompt, type help for some help.

To exit the shell, type ^D.
Psy Shell v0.8.17 (PHP 7.1.20 — cli) by Justin Hileman
New version is available (current: v0.8.17, latest: v0.9.8)
>>> $mediagalleryProcessor = $di->get(\Magento\CatalogImportExport\Model\Import\Product\MediaGalleryProcessor::class)
=> Magento\CatalogImportExport\Model\Import\Product\MediaGalleryProcessor {#10307
     +skuProcessor: Magento\CatalogImportExport\Model\Import\Product\SkuProcessor {#10293},
   }
>>> $di->get(\Magento\Framework\App\State::class)->setAreaCode('frontend')
=> null
>>> $product = $di->get(\Magento\CatalogImportExport\Model\Import\Product::class)
=> Magento\CatalogImportExport\Model\Import\Product {#13163
     +skuProcessor: Magento\CatalogImportExport\Model\Import\Product\SkuProcessor {#10293},
   }
>>> 
```

**Comment by dmanners** [comment]
Hi @peterjaap thanks for this report. I think this is a similar issue to https://github.com/magento-engcom/import-export-improvements/issues/47

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96011`, `MAGETWO-96012` were created

**Comment by magento-engcom-team** [comment]
@peterjaap, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.3-develop` branch. Thank you for collaboration.

**Comment by peterjaap** [comment]
@norgeindian nope unfortunately not :(

**Comment by norgeindian** [comment]
@peterjaap , in the end, we finally found the issue. For the import, we created our own console command. If you run it with `n98-magerun`, you run into the `skuProcessor` issue. Running it with default `bin/magento` solves it.
Don't think, that this was the same in your case, but found it worth to mention.

**Comment by cmuench** [comment]
@sprankhub Sorry, I am still stopped by my bicycle accident.
So far we initialize n98-magerun a bit different to the bin/magento.
We have also sometimes issues in the way the config is loaded. I guess it's related. Maybe we can sync the init process of n98-magerun with bin/magento to fix such issues.
